citations-tool/tool/src/java/org/sakaiproject/citation/tool/CitationHelperAction.java

```diff
@@ -979,6 +991,137 @@ protected void doPostJsonResponse(ParameterParser params,
 		return results;
 	}
 
+	protected Map<String, Object> addSection(ParameterParser params, SessionState state) {
+		String message;
+		Map<String, Object> results = new HashMap<String, Object>();
+		try {
+			int locationId = params.getInt("locationId");
+			CitationCollectionOrder.SectionType sectionType = CitationCollectionOrder.SectionType.valueOf(params.getString("sectionType"));
+			CitationCollection collection = getCitationCollection(state, false);
+
+			CitationCollectionOrder citationCollectionOrder = new CitationCollectionOrder(collection.getId(), locationId, sectionType, H1_SECTION_TITLE);
+			getCitationService().saveSection(citationCollectionOrder);
+			message = rb.getString("resource.updated");
+		}
+		catch (Exception e){
+			message = e.getMessage();
+			logger.warn("Exception in addSection() " + e);
+		}
+		if(message != null && ! message.trim().equals("")) {
+			results.put("message", message);
+		}
+		return results;
+	}
+
+	protected Map<String, Object> addSubSection(ParameterParser params, SessionState state) {
+		String message;
+		Map<String, Object> results = new HashMap<String, Object>();
+
+		try {
+			int locationId = params.getInt("locationId");
+			String addSectionHTML = params.getString("addSectionHTML");
+			CitationCollectionOrder.SectionType sectionType = CitationCollectionOrder.SectionType.valueOf(params.getString("sectionType"));
+			CitationCollection collection = getCitationCollection(state, false);
+
+			CitationCollectionOrder citationCollectionOrder = new CitationCollectionOrder(collection.getId(), locationId, sectionType, addSectionHTML);
+			getCitationService().saveSubsection(citationCollectionOrder);
+			message = rb.getString("resource.updated");
+		}
+		catch (Exception e){
+			message = e.getMessage();
+			logger.warn("Exception in addSubSection() " + e);
+		}
+
+		if(message != null && ! message.trim().equals("")) {
+			results.put("message", message);
+		}
+		return results;
+	}
+
+	protected Map<String, Object> dragAndDrop(ParameterParser params, SessionState state) {
+
+		String message = null;
+		Map<String, Object> results = new HashMap<String, Object>();
+		ObjectMapper mapper = new ObjectMapper();
+		List<CitationCollectionOrder> citationCollectionOrders;
+
+		try {
+			String citationCollectionId = params.getString("citationCollectionId");
+			String nestedCitations = params.getString("data");
+			if (nestedCitations!=null){
+
+				// remove extra parentheses in json
+				// needed because of the extra ol's for accordion effect
+				nestedCitations = nestedCitations
```

Comment:
Yes this JSON parsing is a little brittle though it's relatively unlikely that the HTML layout will change in such a way that it will affect it - mainly because the parsing relies on the HTML structure being a certain way that isn't likely to change, namely ol's (representing the accordions) with li's under them (representing each section/citation) like this:

<ol>
   <li>
        <ol></ol>
   </li>
   <li>...</li>
   <li>...</li>

What is more brittle is that quite a lot of the functionality relies on an element having a particular id or class, e.g., to edit the correct value of an h1 section, that section needs to have the same id as the corresponding row's location in the db.  Though I'm not sure what better ways there are for doing this.